contracts/polyx-vesting/lib.rs

```diff
@@ -0,0 +1,168 @@
+        #[ink(message)]
+        pub fn released(&self) -> Balance {
+            self.released
+        }
+
+        /// Returns the amount of releasable POLYX.
+        #[ink(message)]
+        pub fn releasable(&self) -> Balance {
+            self.vested_amount(self.env().block_timestamp().into())
+                .saturating_sub(self.released())
+        }
+
+        #[ink(message)]
+        pub fn get_time(&self) -> Balance {
+            self.env().block_timestamp().into()
+        }
+
+        #[ink(message)]
+        pub fn get_balance(&self) -> Balance {
+            self.env().balance()
+        }
+
+        /// Release the native token (POLYX) that have already vested.
+        #[ink(message)]
+        pub fn release(&mut self) -> Result<()> {
+            let amount = self.releasable();
+            assert!(amount > 0, "insufficient funds!");
+            self.released += amount;
+            Self::env().emit_event(PolyxReleased { value: amount });
+            if self.env().transfer(self.env().caller(), amount).is_err() {
+                Err(Error::InsufficientBalance)
+            } else {
+                Ok(())
+            }
+        }
+
+        /// Calculates the amount of tokens that has already vested.
+        #[ink(message)]
+        pub fn vested_amount(&self, timestamp: u128) -> Balance {
+            self.vesting_schedule(self.env().balance() + self.released, timestamp)
+        }
+
+        /// This returns the amount vested.
+        fn vesting_schedule(&self, total_allocation: u128, timestamp: u128) -> Balance {
+            if timestamp < self.start {
+                return 0;
+            } else if timestamp > self.start + self.duration {
+                return total_allocation;
+            } else {
+                return (total_allocation * (timestamp - self.start)) / self.duration;
+            }
+        }
+    }
+
+    #[cfg(test)]
+    mod tests {
+        /// Imports all the definitions from the outer scope so we can use them here.
+        use super::*;
+
+        /// Imports `ink_lang` so we can use `#[ink::test]`.
+        use ink_lang as ink;
+
+        /// We test if the new constructor does its job.
+        #[ink::test]
+        fn new_works() {
+            // Constructor works.
+            let polyx_vesting = PolyxVesting::new(AccountId::from([0x01; 32]), 5, 20);
+            // Ensure the values are stored correctly
+            assert_eq!(polyx_vesting.beneficiary(), AccountId::from([0x01; 32]));
+            assert_eq!(polyx_vesting.start(), 5u128);
+            assert_eq!(polyx_vesting.duration(), 20u128);
+            assert_eq!(polyx_vesting.released(), 0u128);
+        }
+
+        #[ink::test]
+        fn vesting_works() {
+            // Constructor works.
+            let mut polyx_vesting = PolyxVesting::new(AccountId::from([0x01; 32]), 1, 3);
+
+            // Release Polyx
+            assert_eq!(polyx_vesting.release(), Ok(()));
```

Comment:
The vesting schedule uses the block timestamp, to calculate how much is releasable.  Right here the timestamp will be zero, since no blocks have been "produced", so the releasable amount is zero.  Need to advance to the next block to release some POLYX in the vesting schedule.

There are Ink! testing methods that can be used to get the contract address, advance to the next block, set account balances, etc...: https://docs.rs/ink_env/latest/ink_env/test/index.html

I good example of Ink! contract testing can be seen here: https://github.com/paritytech/ink/blob/v3.4.0/examples/erc20/lib.rs

Please expand this test to check:

1. Calling `release()` before the start of the vesting period returns an error (right now it panics).
2. Calling `release()` after advancing past the start will release funds.  Also check the beneficiary account balance before and after to make sure they receive the correct funds.
3. Calling `release()` from a non-beneficiary account, still only releases funds to the beneficiary and not the caller.
4. Advance the block timestamp into the next vesting period to release more funds and call `release()` again.

Reply:
Also note that the `block_timestamp` during testing start at zero and increases by `6` each time the chain is advanced (`ink_env::test::advance_block`).

To test the vesting schedule it would be better to use larger `start` and `duration` numbers.

Reply:
working on these changes now.